File: frontend/lib/repositories/item_repository.dart
import 'dart:io';
import 'package:grocery_list/services/api/group_api_client.dart';
import 'package:hive/hive.dart';
import '../models/group.dart';
import '../models/item.dart';
import '../services/api/item_api_client.dart';

class ItemRepository {
  final ItemApiClient _api;
  final GroupApiClient _grApi;
  final Box<GroceryGroup> _groupBox = Hive.box<GroceryGroup>('groups');

  ItemRepository(this._api, this._grApi);

  Future<bool> _isShared(String groupId) async {
    print("DEBUG: Entrée dans _isShared pour $groupId");
    if (groupId == 'default') {
      return false;
    }
    var group = _groupBox.get(groupId);
    group ??= await _grApi.fetchGroup(groupId);
    return group.isShared;
  }

  /// GET: Absolute truth from server
  Future<List<GroceryItem>> getItems(String listId, String groupId) async {
    if (await _isShared(groupId)) {
      return await _api.fetchItems(listId);
    }
    return []; // Logic for local groups can be added here later
  }

  /// POST: Creates item on server and gets the Real ID
  Future<void> addItemToList({
    required String name,
    required String listId,
    required String groupId,
    String? note,
    File? imageFile,
  }) async {
    if (!await _isShared(groupId)) return;

    String? finalImagePath;
    if (imageFile != null) {
      finalImagePath = await _api.uploadImage(imageFile);
    }

    final newItem = GroceryItem(
      name: name,
      status: ItemStatus.pending,
      createdAt: DateTime.now(),
      listId: listId,
      groupId: groupId,
      note: note,
      imagePath: finalImagePath,
    );

    final serverItem = await _api.addItem(newItem);
    newItem.id = serverItem.id;
  }

  /// PUT: Update the item
  Future<void> updateItem(GroceryItem item, String groupId) async {
    if (await _isShared(groupId)) {
      await _api.updateItem(item, groupId);
    }
  }

  Future<void> updateItemDetails({
    required GroceryItem item,
    required String newName,
    String? newNote,
    File? newImageFile,
    bool shouldClearImage = false,
    required String groupId,
  }) async {

    item.name = newName;
    item.note = newNote;

    if (!await _isShared(groupId)) {
      if (shouldClearImage) {
        item.imagePath = null;
      } else if (newImageFile != null) {
        item.imagePath = newImageFile.path;
      }
      await item.save();
      return;
    }

    if (shouldClearImage) {
      item.imagePath = null;
    } else if (newImageFile != null) {
      final String? uploadedPath = await _api.uploadImage(newImageFile);
      if (uploadedPath != null) {
        item.imagePath = uploadedPath;
      }
    }

    await _api.updateItem(item, groupId);
  }

  /// DELETE: Remove from server
  Future<void> deleteItem(GroceryItem item, String groupId) async {
    final bool shared = await _isShared(groupId);

    print("is shared: $shared");
    print("id is not null: ${item.id != null}");

    if (shared && item.id != null) {
      await _api.deleteItem(item.id!, item.name, item.listId, item.groupId);
    }
  }
}

File: frontend_react/src/repositories/itemRepository.ts
import { groupRepository } from './groupRepository';
import {groupApiClient} from "@/services/api/groupApiClient";
import { itemApiClient} from "@/services/api/itemApiClient";
import {GroceryItem, ItemStatus} from "@/types/models";

// Interface defining the cross-platform React Native file references
export interface FilePayload {
    uri: string;
    name: string;
    type: string;
}

class ItemRepository {

    private async _isShared(groupId: string): Promise<boolean> {
        console.log(`DEBUG: Entrée dans _isShared pour ${groupId}`);
        if (groupId === 'default') {
            return false;
        }

        const cachedGroups = await groupRepository.getCachedGroups();
        let group = cachedGroups.find(g => g.id === groupId);

        if (!group) {
            try {
                group = await groupApiClient.fetchGroup(groupId);
            } catch (e) {
                console.error("Repository: Failed to fetch group data remotely", e);
                return false;
            }
        }

        return group ? (group.isShared || false) : false;
    }

    /**
     * GET: Absolute truth from server
     */
    async getItems(listId: string, groupId: string): Promise<GroceryItem[]> {
        if (await this._isShared(groupId)) {
            return await itemApiClient.fetchItems(listId);
        }
        return [];
    }

    /**
     * POST: Creates item on server and gets the Real ID
     */
    async addItemToList({
                            name,
                            listId,
                            groupId,
                            note,
                            imageFile,
                        }: {
        name: string;
        listId: string;
        groupId: string;
        note?: string;
        imageFile?: FilePayload;
    }): Promise<void> {
        if (!(await this._isShared(groupId))) return;

        let finalImagePath: string | undefined;
        if (imageFile) {
            const uploadedPath = await itemApiClient.uploadImage(imageFile as any);
            if (uploadedPath) finalImagePath = uploadedPath;
        }

        const newItem: GroceryItem = {
            createdAt: new Date(),
            status: "pending",
            name,
            listId,
            groupId,
            note
        };

        const serverItem = await itemApiClient.addItem(newItem);
        newItem.id = serverItem.id;
    }

    /**
     * PUT: Update the item
     */
    async updateItem(item: GroceryItem, groupId: string): Promise<void> {
        if (await this._isShared(groupId)) {
            await itemApiClient.updateItem(item, groupId);
        }
    }

    async updateItemDetails({
                                item,
                                newName,
                                newNote,
                                newImageFile,
                                shouldClearImage = false,
                                groupId,
                            }: {
        item: GroceryItem;
        newName: string;
        newNote?: string;
        newImageFile?: FilePayload;
        shouldClearImage?: boolean;
        groupId: string;
    }): Promise<void> {
        item.name = newName;
        item.note = newNote;

        if (!(await this._isShared(groupId))) {
            if (shouldClearImage) {
                item.imagePath = undefined;
            } else if (newImageFile) {
                item.imagePath = newImageFile.uri;
            }
            return;
        }

        if (shouldClearImage) {
            item.imagePath = undefined;
        } else if (newImageFile) {
            const uploadedPath = await itemApiClient.uploadImage(newImageFile as any);
            if (uploadedPath) {
                item.imagePath = uploadedPath;
            }
        }

        await itemApiClient.updateItem(item, groupId);
    }

    /**
     * DELETE: Remove from server
     */
    async deleteItem(item: GroceryItem, groupId: string): Promise<void> {
        const shared = await this._isShared(groupId);
        if (shared && item.id) {
            await itemApiClient.deleteItem(item.id, item.name, item.listId, groupId);
        }
    }
}

export const itemRepository = new ItemRepository();